src/persist-client/src/internal/state_diff.rs

```diff
@@ -0,0 +1,576 @@
+                        // frontier is for these batches (nothing in them, so nothing could have been compacted.
+                        Antichain::from_elem(T::minimum()),
+                    ),
+                    keys: vec![],
+                    len: 0,
+                });
+                metrics.state.apply_spine_fast_path.inc();
+                return Ok(());
+            }
+        }
+        // Fall-through
+        _ => {}
+    }
+
+    // Fast-path: compaction
+    //
+    // TODO: This seems to tickle some existing bugs that we haven't been
+    // hitting because, before incremental state, we were reconstructing Spine
+    // from scratch whenever we deserialized a new version of state. Losing it
+    // is unfortunate, but no worse than where we were before and we're getting
+    // down to the wire with getting inc state (and its backward
+    // incompatibility) into the desired release. Revisit as time permits to
+    // shake out the bugs (they repro readily in CI) and re-enable.
+    #[cfg(TODO)]
+    if let Some((_inputs, output)) = sniff_compaction(&diffs) {
+        let res = FueledMergeRes { output };
+        // We can't predict how spine will arrange the batches when it's
+        // hydrated. This means that something that is maintaining a Spine
+        // starting at some seqno may not exactly match something else
+        // maintaining the same spine starting at a different seqno. (Plus,
+        // maybe these aren't even on the same version of the code and we've
+        // changed the spine logic.) Because apply_merge_res is strict,
+        // we're not _guaranteed_ that we can apply a compaction response
+        // that was generated elsewhere. Most of the time we can, though, so
+        // count the good ones and fall back to the slow path below when we
+        // can't.
+        if trace.apply_merge_res(&res) {
+            // Maybe return the replaced batches from apply_merge_res and verify
+            // that they match _inputs?
+            metrics.state.apply_spine_fast_path.inc();
+            return Ok(());
+        }
+    }
+
+    // Something complicated is going on, so reconstruct the Trace from scratch.
+    metrics.state.apply_spine_slow_path.inc();
+    debug!(
+        "apply_diffs_spine didn't hit a fast-path diffs={:?} trace={:?}",
+        diffs, trace
+    );
+
+    let mut batches = BTreeMap::new();
+    trace.map_batches(|b| assert!(batches.insert(b.clone(), ()).is_none()));
+    apply_diffs_map("spine", diffs, &mut batches)?;
+
+    let mut new_trace = Trace::default();
+    new_trace.downgrade_since(trace.since());
+    for (batch, ()) in batches {
+        // Ignore merge_reqs because whichever process generated this diff is
+        // assigned the work.
+        let _merge_reqs = new_trace.push_batch(batch);
+    }
+    *trace = new_trace;
+    Ok(())
+}
+
+#[allow(dead_code)]
+fn sniff_compaction<'a, T: Timestamp + Lattice>(
```

Comment:
for me, that feels a bit too much in the "log of commands" direction, but this this method has come up with question from you in both of your review passes, which I think is a signal worth paying attention to

I don't plan on re-enabling sniff_compaction before merge (falling back to the slow path for compaction feels quite a bit less risky given that's the path that's always used pre-inc-state, and this PR wants any risk reduction it can get), so left resolving how to do this as a TODO